workspace/__ardulib__/WS2812_NoBufffer/WS2812.cpp

```diff
@@ -0,0 +1,152 @@
+#include "WS2812.h"
+
+WS2812::WS2812(uint16_t pin, uint32_t len) {
+    _length = len;
+    setup(pin);
+}
+
+int8_t getBitPosition(uint8_t b) {
+    int8_t pos = -1;
+    uint8_t t = b;
+    while (t > 0) {
+        t >>= 1;
+        pos++;
+    }
+    return pos;
+}
+
+// Set the specified pin up as digital out
+void WS2812::setup(uint16_t pin) {
+    pinMode(pin, OUTPUT);
+    digitalWrite(pin, LOW);
+    _port = portOutputRegister(digitalPinToPort(pin));
+#if defined(ESP8266) || defined(ESP32)
+    _bit = pin;
+#else
+    _bit = getBitPosition(digitalPinToBitMask(pin));
+#endif
+}
+
+uint32_t WS2812::getLength() {
+    return _length;
+}
+
+#define SEND_BIT(_ON, _OFF, _PORT, _BIT)              \
+    if (_port == &_PORT && _bit == _BIT) {            \
+        asm volatile(                                 \
+            "sbi %[port], %[bit] \n\t"                \
+            ".rept %[onCycles] \n\t"                  \
+            "nop \n\t"                                \
+            ".endr \n\t"                              \
+            "cbi %[port], %[bit] \n\t"                \
+            ".rept %[offCycles] \n\t"                 \
+            "nop \n\t"                                \
+            ".endr \n\t" ::                           \
+                [port] "I"(_SFR_IO_ADDR(_PORT)),      \
+            [bit] "I"(_BIT),                          \
+            [onCycles] "I"(NS_TO_CYCLES(_ON) - 2),    \
+            [offCycles] "I"(NS_TO_CYCLES(_OFF) - 2)); \
+    }
+
+#define SEND_BITS(_ON, _OFF, _PORT) \
+    SEND_BIT(_ON, _OFF, _PORT, 0);  \
+    SEND_BIT(_ON, _OFF, _PORT, 1);  \
+    SEND_BIT(_ON, _OFF, _PORT, 2);  \
+    SEND_BIT(_ON, _OFF, _PORT, 3);  \
+    SEND_BIT(_ON, _OFF, _PORT, 4);  \
+    SEND_BIT(_ON, _OFF, _PORT, 5);
```

Comment:
Why 6, not 8? :thinking:

Reply:
Oops. Overlooked that the PORTB register has only six accessible bits and forgot about other ones. Fixed it.